lib/collection/src/collection/query.rs

```diff
@@ -73,6 +140,22 @@ impl Collection {
         let shard_holder = self.shards_holder.read().await;
         let target_shards = shard_holder.select_shards(shard_selection)?;
 
+        let num_unique_shard_keys = target_shards
+            .iter()
+            .map(|(_, shard_key)| shard_key)
+            .unique()
+            .count();
+        // Auto-sharding happens when we are only querying shards with _the_same_ shard key.
+        // It either might be when we are querying a specific shard key
+        // OR when we are querying all shards with no shard keys specified.
+        let is_auto_sharding = num_unique_shard_keys == 1;
+
+        let batch_request = Self::modify_shard_query_for_undersampling_limits(
+            batch_request,
+            target_shards.len(),
+            is_auto_sharding,
+        );
```

Comment:
I was trying to find a better place for this modification, but was not able to, because:

- Higher level functions require to know original limit and offset of the query for proper merge operation, so we can't modify original query higher in the call stack
- Modification of the limit lower in the call stack also problematic, as we need to know "context" which depends on the shard selection. And we should not know about shard selector inside a shard.